File: src/main/java/com/matag/game/turn/action/shuffle/ShuffleTargetGraveyardIntoLibraryAction.java
package com.matag.game.turn.action.shuffle;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import com.matag.game.cardinstance.CardInstance;
import com.matag.cards.ability.AbilityAction;
import com.matag.cards.ability.Ability;
import com.matag.game.status.GameStatus;

@Component
public class ShuffleTargetGraveyardIntoLibraryAction implements AbilityAction {
  private static final Logger LOGGER = LoggerFactory.getLogger(ShuffleTargetGraveyardIntoLibraryAction.class);

  @Override
  public void perform(CardInstance cardInstance, GameStatus gameStatus, Ability ability) {
    var targetPlayerName = (String) cardInstance.getModifiers().getTargets().get(0);
    var targetPlayer = gameStatus.getPlayerByName(targetPlayerName);

    var graveyardCards = targetPlayer.getGraveyard().extractAllCards();
    targetPlayer.getLibrary().addCards(graveyardCards);
    targetPlayer.getLibrary().shuffle();

    LOGGER.info("{} drew shuffled graveyard into library.", targetPlayer.getName());
  }
}

File: src/main/kotlin/com/matag/game/turn/action/shuffle/ShuffleTargetGraveyardIntoLibraryAction.kt
package com.matag.game.turn.action.shuffle

import com.matag.cards.ability.Ability
import com.matag.cards.ability.AbilityAction
import com.matag.game.cardinstance.CardInstance
import com.matag.game.status.GameStatus
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class ShuffleTargetGraveyardIntoLibraryAction : AbilityAction {
    override fun perform(cardInstance: CardInstance, gameStatus: GameStatus, ability: Ability?) {
        val targetPlayerName = cardInstance.modifiers.targets.get(0) as String?
        val targetPlayer = gameStatus.getPlayerByName(targetPlayerName)

        val graveyardCards = targetPlayer.graveyard.extractAllCards()
        targetPlayer.library.addCards(graveyardCards)
        targetPlayer.library.shuffle()

        LOGGER.info("{} drew shuffled graveyard into library.", targetPlayer.name)
    }

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(ShuffleTargetGraveyardIntoLibraryAction::class.java)
    }
}
